File: 54-spiral-matrix/spiral-matrix.cpp
#define add push_back

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& arr) {
        int n=arr.size(),m=arr[0].size();
        vector<int> lis;
    int left=0,right=m-1;
    int top=0,btm=n-1;
    int dir=0;
    while(top<=btm && left<=right){
        if(dir==0){
            for(int x=left;x<=right;++x){
                //                     System.out.println(arr[top][x]);
                lis.add(arr[top][x]);
            }
            ++top;
        }
        else if(dir==1){
            for(int x=top;x<=btm;++x){
                //                     System.out.println(arr[x][right]);
                lis.add(arr[x][right]);
            }
            --right;
        }
        else if(dir==2){
            for(int x=right;x>=left;--x){
                //                     System.out.println(arr[btm][x]);
                lis.add(arr[btm][x]);
            }
            --btm;
        }
        else if(dir==3){
            for(int x=btm;x>=top;--x){
                //                     System.out.println(arr[x][left]);
                lis.add(arr[x][left]);
            }
            ++left;
        }
        dir=(dir+1)%4;
    }
    return lis;
    }
};

File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] A) {
        List<Integer> res = new ArrayList<>();
        int top=0, btm=A.length-1, lft=0, rgt= A[0].length-1, dir=0;
        while(top<=btm && lft<=rgt){
            if(dir==0){
                for(int i=lft;i<=rgt;i++) res.add(A[top][i]);
                top++;
            }
            else if(dir==1){
                for(int i=top;i<=btm;i++) res.add(A[i][rgt]);
                rgt--;
            }
            else if(dir==2){
                for(int i=rgt;i>=lft;i--) res.add(A[btm][i]);
                btm--;
            }
            else{
                for(int i=btm;i>=top;i--) res.add(A[i][lft]);
                lft++;
            }
            dir= (dir+1)%4;
        }
        return res;
    }
}